### PR TITLE
[FS-118429] : Background refill mechanism for SW cached modules in FS

### PR DESCRIPTION
**Pull request by preethi-sathiyanarayanan**
## Changes proposed in this pull request
SW - CLEAR_AND_REFILL_API_CACHE - Method will receive the list of URLs to delete from cache and make a background fetch call to refill with updated data. Post this a callback will be triggered back to the callee with an uniqueId(triggeredFrom - which was sent from the callee)

### CODE REVIEW

File: service-worker/index.js [hunk 1]
```diff
@@ -20,12 +20,19 @@ const FETCH_DATA = (event, cacheName) => {
     caches.open(cacheName).then((cache) => {
       return cache.match(request).then((response) => {
         if (response) {
+          console.log('SW::FETCH_DATA:: SW has response in cache');
```
Comment (preethi-sathiyanarayanan) [inline service-worker/index.js]: will remove unnecessary console statements post testing

File: service-worker/index.js [hunk 2]
```diff
@@ -49,34 +76,41 @@ const POST_MSG_TO_ALL_CLIENTS = (clients, event) => {
...
-        if (response.status == 200) {
-          cache.put(request, response.clone());
+        if(response.status == 200) {
+          let clonedResp = response.clone();
+
+          let modifiedHeaders = new Headers([...clonedResp.headers, ['from-sw', true]]);
+          let updatedResponse = new Response(clonedResp.body, {headers: modifiedHeaders});
+
+          console.log(`SW::CUSTOM_FETCH:: triggered cache Put with modified headers => updatedResponse :${updatedResponse}`);
+          cache.put(request, updatedResponse);
```
Comment (vigneshk-fd) [inline service-worker/index.js]: @preethi-sathiyanarayanan :  Please extract this entire if block, into a separate method and reuse in lines 28-34, , 59-66 & 106-113

File: service-worker/index.js [hunk 3]
```diff
@@ -99,12 +133,14 @@ const CUSTOM_FETCH = (event) => {
 
 // Force put to sw-cache
 const CUSTOM_PUT = (event) => {
-  console.log(`SW:: Custom sw-cache-put has been triggered for '${event.data.url}'`);
+  console.log(`SW::CUSTOM_PUT:: Custom sw-cache-put has been triggered for '${event.data.url}'`);
 
   let request = new Request(event.data.url, event.data.options);
   
+  let modifiedHeaders = new Headers([...event.data.options.headers, ['from-sw', true]]);
```
Comment (vigneshk-fd) [inline service-worker/index.js]: Also, let's keep the setting of `from-sw: true` in the response headers in a single place via single method.

**Review by vigneshk-fd — CHANGES_REQUESTED** [review]
Changes requested

File: service-worker/index.js (hunk 3 above)
Comment (preethi-sathiyanarayanan) [reply service-worker/index.js]: This method has a different pattern for response build, hence unable to move it to the common function(CACHE_PUT_TO_SW).

File: service-worker/index.js [hunk 4]
```diff
@@ -15,134 +15,165 @@ const API_PATTERN_REGEX = API_PATTERNS.map(createRegEx);
...
+			return cache.match(request).then((response) => {
+				if (response) {
+					console.log('SW::FETCH_DATA:: SW has response in cache for : ', request.url);
+					return response;
+				}
+				console.log('SW::FETCH_DATA:: sw has no cache. Triggering a fetch call for :', request.url);
+				return fetch(request).then((response) => {
+					if(response.status == 200) {
+						let clonedResp = response.clone();
+						CACHE_PUT_TO_SW(cache, request, clonedResp);
```
Comment (vigneshk-fd) [inline service-worker/index.js]:
```suggestion
						CACHE_PUT_TO_SW(cache, request, response.clone());
```

File: service-worker/index.js [hunk 5]
```diff
@@ -15,134 +15,165 @@ const API_PATTERN_REGEX = API_PATTERNS.map(createRegEx);
...
-        caches.delete(cacheName);
-      }
-    });
-  });
+const CLEAR_AND_REFILL_API_CACHE = (event) => {
+	let options = event.data.options;
+	let urlListToCacheReset = event.data.urlListToCacheReset || [];
+	let sourceClient =  event.source;
+	console.log('SW::CLEAR_AND_REFILL_API_CACHE:: addon going to clear api cache for APIs : ', urlListToCacheReset);
+	caches.open(API_CACHE_NAME).then((cache) => {
```
Comment (vigneshk-fd) [inline service-worker/index.js:45]:
```suggestion
	
	caches.open(API_CACHE_NAME).then((cache) => {
```

File: service-worker/index.js [hunk 6]
```diff
@@ -15,134 +15,165 @@ const API_PATTERN_REGEX = API_PATTERNS.map(createRegEx);
...
+			cache.delete(url).then(() => {
+				console.log('SW::CLEAR_AND_REFILL_API_CACHE:: deleted SW cache for url : ', url);
+				let request = new Request(url, options);
+				let headers = options.headers || {};
+
+				console.log('SW::CLEAR_AND_REFILL_API_CACHE:: Fetch call trigger for url : ', url);
+				fetch(request, { headers }).then((response) => {
+					console.log(`SW::CLEAR_AND_REFILL_API_CACHE:: Fetch call completed for url :${url} with response :${response}`);
+					if(response.status == 200) {
+						CACHE_PUT_TO_SW(cache, request, response);
```
Comment (vigneshk-fd) [inline service-worker/index.js:56]:
```suggestion
						CACHE_PUT_TO_SW(cache, request, response.clone());
```

File: service-worker/index.js [hunk 7]
```diff
@@ -15,134 +15,165 @@ const API_PATTERN_REGEX = API_PATTERNS.map(createRegEx);
...
+		return cache.match(request).then((response) => {
+			if (response) {
+				console.log('SW::CUSTOM_FETCH:: sw has response in cache');
+				return response;
+			}
+			// Fallback !
+			return fetch(request, { headers: event.data.options.headers || {}}).then((response) => {
+				if(response.status == 200) {
+					let clonedResp = response.clone();
+					CACHE_PUT_TO_SW(cache, request, clonedResp);
```
Comment (vigneshk-fd) [inline service-worker/index.js]:
```suggestion
					CACHE_PUT_TO_SW(cache, request, response.clone());
```

File: service-worker/index.js [hunk 8]
```diff
@@ -15,134 +15,165 @@ const API_PATTERN_REGEX = API_PATTERNS.map(createRegEx);
...
 };
 
+const CACHE_PUT_TO_SW = (cache, request, response) => {
+	let modifiedHeaders = new Headers([...response.headers, ['from-sw', true]]);
+	let updatedResponse = new Response(response.body, {headers: modifiedHeaders});
+
+	console.log(`SW::cachePutToSW:: triggered cache Put with modified headers => updatedResponse :${updatedResponse}`);
+	cache.put(request, updatedResponse);
+}
+
```
Comment (vigneshk-fd) [inline service-worker/index.js:147]: Will this work ?

```suggestion
	console.log(`SW::CUSTOM_PUT:: Custom sw-cache-put has been triggered for '${event.data.url}'`);

	let request = new Request(event.data.url, event.data.options);
	let modifiedHeaders = buildHeaders(event.data.options);

	// event.data.payload --> Payload to push to cache, sent from client
	let response = new Response(event.data.payload, { status: 200, statusText: 'ok', headers: modifiedHeaders}); // blob

	caches.open(API_CACHE_NAME).then((cache) => {
		cache.put(request, response);

		self.clients.matchAll().then((clients) => {
			POST_MSG_TO_ALL_CLIENTS(clients, event);
		});
	});
};

const buildHeaders = ({ headers }) => new Headers([...headers, ['from-sw', true]]);

const CACHE_PUT_TO_SW = (cache, request, response) => {
	let modifiedHeaders = buildHeaders(response.headers);
	let updatedResponse = new Response(response.body, {headers: modifiedHeaders});

	console.log(`SW::cachePutToSW:: triggered cache Put with modified headers => updatedResponse :${updatedResponse}`);
	cache.put(request, updatedResponse);
}

```

File: service-worker/index.js [hunk 9]
```diff
@@ -15,134 +15,165 @@ const API_PATTERN_REGEX = API_PATTERNS.map(createRegEx);
...
-        });
-      })
-    })
-  );
+	let request = event.request;
+	event.respondWith(
+		caches.open(cacheName).then((cache) => {
+			return cache.match(request).then((response) => {
+				if (response) {
+					console.log('SW::FETCH_DATA:: SW has response in cache for : ', request.url);
```
Comment (vigneshk-fd) [inline service-worker/index.js:24]: **General thought** : Once testing is complete in stack and before going live, if possible, pls re-organise the log messages to log important points in execution alone, (with prefixing msg as `FS-SW::`, etc.,) and making it turn on based on flag or so (ex: via query params `sw-logs=true`, etc just like BR logs)

File: service-worker/index.js [hunk 10]
```diff
@@ -15,134 +15,165 @@ const API_PATTERN_REGEX = API_PATTERNS.map(createRegEx);
...
-          if(response.status == 200) {
-            cache.put(request, response.clone());
-          }
-          return response;
-        });
-      })
-    })
-  );
+	let request = event.request;
+	event.respondWith(
```
Comment (vigneshk-fd) [inline service-worker/index.js:20]:
```suggestion

	event.respondWith(
```

File: service-worker/index.js [hunk 11]
```diff
@@ -15,134 +15,165 @@ const API_PATTERN_REGEX = API_PATTERNS.map(createRegEx);
...
-    });
-  });
+	console.log(`SW::CUSTOM_PUT:: Custom sw-cache-put has been triggered for '${event.data.url}'`);
+
+	let request = new Request(event.data.url, event.data.options);
+	
+	let modifiedHeaders = new Headers([...event.data.options.headers, ['from-sw', true]]);
+
+	// event.data.payload --> Payload to push to cache, sent from client
+	let response = new Response(event.data.payload, { status: 200, statusText: 'ok', headers: modifiedHeaders}); // blob
```
Comment (vigneshk-fd) [inline service-worker/index.js]:
```suggestion
	let response = new Response(event.data.payload, { status: 200, statusText: 'ok', headers: modifiedHeaders }); // blob
```

File: service-worker/index.js [hunk 12]
```diff
@@ -15,134 +15,165 @@ const API_PATTERN_REGEX = API_PATTERNS.map(createRegEx);
...
+	});
 };
 
+const CACHE_PUT_TO_SW = (cache, request, response) => {
+	let modifiedHeaders = new Headers([...response.headers, ['from-sw', true]]);
+	let updatedResponse = new Response(response.body, {headers: modifiedHeaders});
+
+	console.log(`SW::cachePutToSW:: triggered cache Put with modified headers => updatedResponse :${updatedResponse}`);
+	cache.put(request, updatedResponse);
+}
```
Comment (vigneshk-fd) [inline service-worker/index.js]:
```suggestion

const buildHeaders = ({ headers }) => new Headers([...headers, ['from-sw', true]]);

const CACHE_PUT_TO_SW = (cache, request, response) => {
	let modifiedHeaders = buildHeaders(response);
	let updatedResponse = new Response(response.body, {headers: modifiedHeaders});

	console.log(`SW::cachePutToSW:: triggered cache Put with modified headers => updatedResponse :${updatedResponse}`);
	cache.put(request, updatedResponse);
}
```

File: service-worker/index.js [hunk 13]
```diff
@@ -15,134 +15,165 @@ const API_PATTERN_REGEX = API_PATTERNS.map(createRegEx);
...
-    cache.put(request, response);
-    self.clients.matchAll().then((clients) => {
-      POST_MSG_TO_ALL_CLIENTS(clients, event);
-    });
-  });
+	console.log(`SW::CUSTOM_PUT:: Custom sw-cache-put has been triggered for '${event.data.url}'`);
+
+	let request = new Request(event.data.url, event.data.options);
+	
+	let modifiedHeaders = new Headers([...event.data.options.headers, ['from-sw', true]]);
```
Comment (vigneshk-fd) [inline service-worker/index.js]:
```suggestion
	let modifiedHeaders = builderHeaders(event.data.options);
```

**Review by vigneshk-fd — CHANGES_REQUESTED** [review]
Changes requested.

**Review by vigneshk-fd — CHANGES_REQUESTED** [review]
Changes requested.

File: service-worker/index.js (hunk 6 above)
Comment (preethi-sathiyanarayanan) [reply service-worker/index.js:56]: we arent returning the response here, so we can directly use it.

In other areas, we were returning the response as well as using the same to PUT in SW. Response object instance can be used only once.

**Review by vigneshk-fd — APPROVED** [review]

**Pull request merged**